components/feature/pwa/src/main/java/mozilla/components/feature/pwa/intent/FennecWebAppIntentProcessor.kt

```diff
@@ -0,0 +1,86 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.pwa.intent
+
+import android.content.Intent
+import android.content.Intent.FLAG_ACTIVITY_NEW_DOCUMENT
+import androidx.annotation.VisibleForTesting
+import mozilla.components.browser.session.Session
+import mozilla.components.browser.session.Session.Source
+import mozilla.components.browser.session.SessionManager
+import mozilla.components.concept.engine.EngineSession
+import mozilla.components.concept.engine.manifest.WebAppManifest
+import mozilla.components.concept.engine.manifest.WebAppManifestParser
+import mozilla.components.concept.engine.manifest.getOrNull
+import mozilla.components.feature.intent.ext.putSessionId
+import mozilla.components.feature.intent.processing.IntentProcessor
+import mozilla.components.feature.pwa.ext.putWebAppManifest
+import mozilla.components.feature.pwa.ext.toCustomTabConfig
+import mozilla.components.feature.session.SessionUseCases
+import mozilla.components.support.utils.toSafeIntent
+import org.json.JSONObject
+import java.io.File
+
+/**
+ * Legacy processor for Progressive Web App shortcut intents created by Fennec.
+ */
+class FennecWebAppIntentProcessor(
+    private val sessionManager: SessionManager,
+    private val loadUrlUseCase: SessionUseCases.DefaultLoadUrlUseCase
+) : IntentProcessor {
+
+    /**
+     * Returns true if this intent should launch a progressive web app created in Fennec.
+     */
+    override fun matches(intent: Intent) =
+        intent.toSafeIntent().action == ACTION_FENNEC_WEBAPP
+
+    /**
+     * Processes the given [Intent] by creating a [Session] with a corresponding web app manifest.
+     *
+     * A custom tab config is also set so a custom tab toolbar can be shown when the user leaves
+     * the scope defined in the manifest.
+     */
+    override suspend fun process(intent: Intent): Boolean {
+        val safeIntent = intent.toSafeIntent()
+        val url = safeIntent.dataString
+
+        return if (!url.isNullOrEmpty() && matches(intent)) {
+            val webAppManifest = fromFile(safeIntent.getStringExtra(EXTRA_FENNEC_MANIFEST_PATH))
+                ?: return false
```

Comment:
Do you think it would be worthwhile to save the recovered manifest in our storage and in future launches only look at the start URL but use the manifest from our storage (which may have been updated)?

Reply:
Yes, that makes sense. I'll build on top of the manifest update code and add in functionality to "promote" the shortcut as well.